### PR TITLE
Deeper Backbone: 4 Transolver Blocks (n_layers 3→4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current 3-block Transolver backbone may be **depth-limited** — each TransolverBlock applies one round of physics-aware slice attention + FFN. Adding a 4th block increases the number of sequential reasoning steps the model can apply to the point cloud, allowing more complex multi-foil physics reasoning before the SRF heads decode surface pressure.

**Why this might work:**
- The only historically successful improvements have been (1) input feature additions and (2) loss formulation changes. No one has tested whether the backbone itself is a capacity bottleneck.
- Wider SRF (frieren #2283) tests **head** capacity. This tests **backbone** capacity — orthogonal experiments.
- With ~45GB VRAM used on H100 (96GB), there's ample headroom for a larger backbone.
- More attention layers = more physics reasoning steps. Boundary layer interactions, wake propagation, and tandem coupling may benefit from deeper sequential processing.

**Risk assessment:**
- Per-epoch time increases ~20-25% → expect ~125-135 epochs in 180 min (vs ~148 baseline). Still reaches late cosine schedule.
- Slight overfitting risk, but EMA + augmentation + cosine LR should mitigate.
- VRAM increase ~5-8GB (well within H100 headroom).

## Instructions

**Single change**: modify `--n_layers` from 3 to 4. Everything else stays identical to baseline.

### Seed 42
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/deeper-backbone-s42" \
  --wandb_group "deeper-backbone" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 4 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### Seed 73
Same command with `--seed 73` and `--wandb_name "alphonse/deeper-backbone-s73"`.

### What to report
1. 2-seed average for p_in, p_oodc, p_tan, p_re
2. Number of epochs completed and wall clock time (important to understand throughput impact)
3. Peak VRAM usage (to gauge if 5+ blocks are feasible)
4. W&B run IDs

## Baseline (PR #2251, 2-seed avg)

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.891** | < 11.89 |
| **p_oodc** | **7.561** | < 7.56 |
| **p_tan** | **28.118** | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B baseline: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Training command (seed 42)
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/deeper-backbone-s42" \
  --wandb_group "deeper-backbone" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 4 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

Seed 73 same command with `--seed 73` and adjusted `--wandb_name`.

### Metrics

| Metric | s42 | s73 | 2-seed avg | baseline (PR #2251) | Δ |
|--------|-----|-----|-----------|---------------------|---|
| **p_in** | 14.0 | 16.7 | **15.35** | 11.891 | **+29.1%** ⛔ |
| **p_tan** | 29.8 | 29.1 | **29.45** | 28.118 | **+4.7%** ⛔ |
| **p_oodc** | 8.8 | 9.2 | **9.00** | 7.561 | **+19.0%** ⛔ |
| **p_re** | 6.9 | 7.4 | **7.15** | 6.364 | **+12.4%** ⛔ |

Surface MAE Ux/Uy (val_in_dist): s42 = 2.89/0.97, s73 = 2.81/0.94

### W&B run IDs
- Seed 42: `tmityb5d` (alphonse/deeper-backbone-s42)
- Seed 73: `spcn43ml` (alphonse/deeper-backbone-s73)

### Throughput and memory
- **Per-epoch time**: ~90s (vs ~68s baseline with 3 layers) → **+32% slower**
- **Epochs completed**: s42 = 121, s73 = 120 (vs ~148-149 baseline)
- **Peak VRAM**: s42 = 54.8 GB, s73 = 54.9 GB (vs ~45 GB baseline, **+22% VRAM**)
- **Wall clock**: s42 = 181.0 min, s73 = 180.7 min

### What happened — negative result

Adding a 4th Transolver block **significantly hurts all metrics** across both seeds. The p_in regression is especially severe at +29.1% (worse than any other recent experiment).

**Primary failure mode — underfitting due to fewer training epochs**: The 4-layer model runs ~32% slower per epoch, completing only 120-121 epochs vs baseline's ~149. With cosine_T_max=150, both models are cut off at only ~80% of the cosine schedule — the LR never reaches its minimum. At epoch 120, the 3-layer baseline had val_in_dist losses of ~0.31-0.33 (similar to 4-layer's final), but the 3-layer model had 28 more epochs to refine, reaching ~0.28 by epoch 149. The deeper model simply doesn't have enough wall-clock budget to converge.

**Secondary factors:**
- High seed variance on p_in (s42=14.0, s73=16.7) suggests the 4-layer model hasn't converged — still in a noisy regime
- The extra layer adds ~10 GB VRAM overhead but the model parameters aren't being used efficiently with only 120 epochs of training
- Both seeds were still improving at termination (loss still decreasing), confirming the model needed more training time

**The backbone is NOT capacity-limited at 3 layers for this task**: The model can't utilize the extra capacity within the fixed training budget. The 3-layer model is better calibrated to the 180-min wall clock constraint — it completes ~99% of its cosine schedule and converges more completely.

### Suggested follow-ups

1. **Adjust cosine_T_max to match epoch budget**: If retesting 4 layers, set `--cosine_T_max 120` to ensure the cosine schedule completes within the reduced epoch budget. This isolates the capacity question from the schedule mismatch.

2. **Test with a longer wall clock budget**: If training can be extended to 240 min, the 4-layer model would complete ~160 epochs — more than enough to finish the schedule. This would test whether deeper = better when training budget is not a constraint.

3. **Width over depth**: Instead of more layers, test wider attention (e.g., `--slice_num 128` or hidden_dim increase). Width doesn't increase per-epoch time as dramatically as depth, so the model can still converge within budget.

4. **Hybrid approach**: Use 4 layers but with a smaller slice_num (e.g., 64) to keep per-epoch time comparable to baseline. Tests whether more reasoning steps at lower per-step resolution beats fewer steps at higher resolution.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2285 because ADVISOR: Closing — 4-layer backbone severely regresses all metrics (p_in +29.1%, p_oodc +19.0%, p_tan +4.7%, p_re +12.4%).

**Root cause — training budget mismatch.** The 4-layer model runs 32% slower per epoch (90s vs 68s), completing only 121 epochs vs baseline's 149. With cosine_T_max=150, the model is cut off at ~80% of its schedule — LR never reaches minimum. The 3-layer baseline completes ~99% of its cosine schedule and converges much more fully.

**Key insight: the 3-layer backbone is well-calibrated to the 180-min training budget.** Adding depth without proportionally increasing training time is counterproductive. The backbone is not capacity-limited for this task — it's training-budget-limited.

Good throughput analysis. The VRAM (+22%) and timing data are useful for future capacity experiments. Moving you to a fresh direction — a novel loss formulation..
